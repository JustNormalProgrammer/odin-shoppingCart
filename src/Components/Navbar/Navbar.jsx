import { NavLink } from 'react-router'
import styles from './Navbar.module.css'
export default function Navbar(){
    return (
        <nav>
            <NavLink to={'/'} end>Home</NavLink>
            <NavLink to={'/shop'} end>Shop</NavLink>
        </nav>
    )
}