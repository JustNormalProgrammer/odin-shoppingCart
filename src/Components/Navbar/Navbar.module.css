.nav{
    background-color: #121212;
    display: flex;
    justify-content: space-between;
    align-items: center;
}
.nav ul{
    display: flex;
    align-items: center;
    justify-content: flex-end;
    z-index: 1;
}

.link{
    height: 60px;
    width: 100px;
    position: relative;
    font-size: 18px;
    color: #8962f3;
    display: flex;
    align-items: center;
    justify-content: center;
    z-index: 2;
}
.link:after{
    content: '';
    position: absolute;
    height: 4px;
    width: 0%;
    background-color: #8962f3;
    bottom: 0;
    left: 50%;
    transition: 0.2s ease-out;
}
.link:hover:after{
    left: 0;
    width: 100%;
    pointer-events: none;
}
:global(a.active):before {
    content: '';
    z-index: -1;
    position: absolute;
    height: 100%;
    width: 100%;
    background-color: #2f293a;
}
:global(a.active):after{
    width: 100%;
    left: 0;
    background-color: #a190fa;
}