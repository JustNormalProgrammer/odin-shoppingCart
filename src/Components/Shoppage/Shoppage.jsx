import styles from './Shoppage.module.css'
export default function Shoppage(){
    return <h1>Hello Shoppage!</h1>
}