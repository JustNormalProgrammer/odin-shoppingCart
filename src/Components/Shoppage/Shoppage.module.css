.container{
    flex-grow: 1;
    display: flex;
    flex-wrap: wrap;
}
@media(min-width: 1100px){
    .container{
        flex-wrap: nowrap;
    }
}