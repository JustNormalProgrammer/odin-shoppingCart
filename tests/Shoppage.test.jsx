import {
  describe,
  it,
  expect,
  beforeAll,
  afterEach,
  afterAll,
  vi,
} from "vitest";
import { render, waitFor } from "@testing-library/react";
import CartContextProvider from "../src/Contexts/CartContext";
import { QueryClientProvider, QueryClient } from "@tanstack/react-query";
import { MemoryRouter } from "react-router";
import Shoppage from "../src/Components/Shoppage/Shoppage";
import Navbar from "../src/Components/Navbar/Navbar";
import server from "./server";
import userEvent from "@testing-library/user-event";
import PropTypes from "prop-types";

Object.defineProperty(window.SVGElement.prototype, "getBBox", {
  writable: true,
  value: vi.fn().mockReturnValue({
    x: 0,
    y: 0,
  }),
});

beforeAll(() => server.listen());

afterEach(() => server.resetHandlers());

afterAll(() => server.close());

const queryClient = new QueryClient();

const Wrapper = ({ children }) => {
  return (
    <QueryClientProvider client={queryClient}>
      <CartContextProvider>{children}</CartContextProvider>
    </QueryClientProvider>
  );
};
Wrapper.propTypes = {
  children: PropTypes.any,
}

describe("Shoppage component", () => {
  it("renders shoppage component with skeleton loaders", () => {
    const { container } = render(<Shoppage />, { wrapper: Wrapper });
    expect(container).toMatchSnapshot();
  });
  it("renders categories of shop items when data fetching is completed", async () => {
    const { findByText } = render(<Shoppage />, { wrapper: Wrapper });
    expect(await findByText("electronics")).toBeInTheDocument();
    expect(await findByText("jewelery")).toBeInTheDocument();
    expect(await findByText("men's clothing")).toBeInTheDocument();
    expect(await findByText("women's clothing")).toBeInTheDocument();
  });
  it("renders shop items", async () => {
    const { findByText } = render(<Shoppage />, { wrapper: Wrapper });
    expect(
      await findByText("Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops")
    ).toBeInTheDocument();
    expect(
      await findByText("Mens Casual Premium Slim Fit T-Shirts")
    ).toBeInTheDocument();
    expect(await findByText("Mens Cotton Jacket")).toBeInTheDocument();
  });
  it("renders only shop items matching selected category", async () => {
    const user = userEvent.setup();
    const { findByText, getByText, queryByText } = render(<Shoppage />, {
      wrapper: Wrapper,
    });
    const categoryButton = await findByText("jewelery");
    await user.click(categoryButton);
    expect(
      getByText(
        "John Hardy Women's Legends Naga Gold & Silver Dragon Station Chain Bracelet"
      )
    ).toBeInTheDocument();
    expect(
      queryByText("Mens Casual Premium Slim Fit T-Shirts")
    ).not.toBeInTheDocument();
    expect(
      queryByText(
        "WD 4TB Gaming Drive Works with Playstation 4 Portable External Hard Drive"
      )
    ).not.toBeInTheDocument();
    expect(
      queryByText("Rain Jacket Women Windbreaker Striped Climbing Raincoats")
    ).not.toBeInTheDocument();
  });
  it("renders input field with initial value of quantity of 1", async () => {
    const { getAllByDisplayValue, findAllByAltText } = render(<Shoppage />, {
      wrapper: Wrapper,
    });
    await findAllByAltText("Add item")[0];
    const inputField = getAllByDisplayValue(1)[0];
    expect(inputField).toBeVisible();
    expect(inputField).toHaveValue("1");
  });
  it("changes the value of quantity field after user click", async () => {
    const user = userEvent.setup();
    const { getAllByAltText, getAllByDisplayValue } = render(<Shoppage />, {
      wrapper: Wrapper,
    });
    await waitFor(async () => {
      expect(getAllByAltText("Add item")[0]).toBeVisible();
    });
    const addItemBtn = getAllByAltText("Add item")[1];
    await user.click(addItemBtn);
    expect(getAllByDisplayValue(2)[0]).toBeInTheDocument();
  });
  it("changes the value of quantity of the cart in the navbar when user adds item to the cart", async () => {
    const user = userEvent.setup();
    const { getAllByAltText, getByText} = render(
      <MemoryRouter>
        <QueryClientProvider client={queryClient}>
          <CartContextProvider>
            <Navbar/>
            <Shoppage />
          </CartContextProvider>
        </QueryClientProvider>
      </MemoryRouter>
    );
    await waitFor(async () => {
      expect(getAllByAltText("Add item")[0]).toBeVisible();
    });
    const addtoCartButton = getAllByAltText("Add to cart")[0];
    expect(getByText("0")).toBeInTheDocument();
    await user.click(addtoCartButton);
    expect(getByText("1")).toBeInTheDocument();
    await user.click(addtoCartButton);
    expect(getByText("2")).toBeInTheDocument();
  });
});
