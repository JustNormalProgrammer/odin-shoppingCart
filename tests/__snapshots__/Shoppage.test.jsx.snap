// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Shoppage component > renders shoppage component with skeleton loaders 1`] = `
<div>
  <div
    class="_container_35a4d7"
  >
    <nav
      class="_navContainer_24044f"
    >
      <div
        class="_navWrapper_24044f"
      >
        <div
          class="_navCard_24044f"
        >
          <div
            class="_sectionHeader_24044f"
          >
            Categories
          </div>
          <div
            class="_cardContent_24044f"
          >
            <span
              aria-busy="true"
              aria-live="polite"
            >
              <span
                class="react-loading-skeleton"
                style="height: 25px; --base-color: #2f293a; --highlight-color: #a190fa;"
              >
                ‌
              </span>
              <br />
              <span
                class="react-loading-skeleton"
                style="height: 25px; --base-color: #2f293a; --highlight-color: #a190fa;"
              >
                ‌
              </span>
              <br />
              <span
                class="react-loading-skeleton"
                style="height: 25px; --base-color: #2f293a; --highlight-color: #a190fa;"
              >
                ‌
              </span>
              <br />
              <span
                class="react-loading-skeleton"
                style="height: 25px; --base-color: #2f293a; --highlight-color: #a190fa;"
              >
                ‌
              </span>
              <br />
            </span>
          </div>
        </div>
        <div
          class="_navCard_24044f"
        >
          <div
            class="_sectionHeader_24044f"
          >
            Sort by
          </div>
          <div
            class="_cardContent_24044f"
          >
            <div
              class=""
            >
              Price asc
            </div>
            <div
              class=""
            >
              Price desc
            </div>
            <div
              class=""
            >
              Rating asc
            </div>
            <div
              class=""
            >
              Rating desc
            </div>
          </div>
        </div>
      </div>
    </nav>
    <div
      class="_shopContainer_19f3cb"
    >
      <div
        class="_shopItem_19f3cb"
      >
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 220px; height: 150px; --base-color: #2f293a; --highlight-color: #a190fa;"
          >
            ‌
          </span>
          <br />
        </span>
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 220px; height: 60px; --base-color: #2f293a; --highlight-color: #a190fa;"
          >
            ‌
          </span>
          <br />
        </span>
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 220px; height: 40px; --base-color: #2f293a; --highlight-color: #a190fa;"
          >
            ‌
          </span>
          <br />
        </span>
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 220px; height: 60px; --base-color: #2f293a; --highlight-color: #a190fa;"
          >
            ‌
          </span>
          <br />
        </span>
      </div>
      <div
        class="_shopItem_19f3cb"
      >
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 220px; height: 150px; --base-color: #2f293a; --highlight-color: #a190fa;"
          >
            ‌
          </span>
          <br />
        </span>
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 220px; height: 60px; --base-color: #2f293a; --highlight-color: #a190fa;"
          >
            ‌
          </span>
          <br />
        </span>
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 220px; height: 40px; --base-color: #2f293a; --highlight-color: #a190fa;"
          >
            ‌
          </span>
          <br />
        </span>
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 220px; height: 60px; --base-color: #2f293a; --highlight-color: #a190fa;"
          >
            ‌
          </span>
          <br />
        </span>
      </div>
      <div
        class="_shopItem_19f3cb"
      >
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 220px; height: 150px; --base-color: #2f293a; --highlight-color: #a190fa;"
          >
            ‌
          </span>
          <br />
        </span>
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 220px; height: 60px; --base-color: #2f293a; --highlight-color: #a190fa;"
          >
            ‌
          </span>
          <br />
        </span>
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 220px; height: 40px; --base-color: #2f293a; --highlight-color: #a190fa;"
          >
            ‌
          </span>
          <br />
        </span>
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 220px; height: 60px; --base-color: #2f293a; --highlight-color: #a190fa;"
          >
            ‌
          </span>
          <br />
        </span>
      </div>
      <div
        class="_shopItem_19f3cb"
      >
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 220px; height: 150px; --base-color: #2f293a; --highlight-color: #a190fa;"
          >
            ‌
          </span>
          <br />
        </span>
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 220px; height: 60px; --base-color: #2f293a; --highlight-color: #a190fa;"
          >
            ‌
          </span>
          <br />
        </span>
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 220px; height: 40px; --base-color: #2f293a; --highlight-color: #a190fa;"
          >
            ‌
          </span>
          <br />
        </span>
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 220px; height: 60px; --base-color: #2f293a; --highlight-color: #a190fa;"
          >
            ‌
          </span>
          <br />
        </span>
      </div>
      <div
        class="_shopItem_19f3cb"
      >
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 220px; height: 150px; --base-color: #2f293a; --highlight-color: #a190fa;"
          >
            ‌
          </span>
          <br />
        </span>
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 220px; height: 60px; --base-color: #2f293a; --highlight-color: #a190fa;"
          >
            ‌
          </span>
          <br />
        </span>
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 220px; height: 40px; --base-color: #2f293a; --highlight-color: #a190fa;"
          >
            ‌
          </span>
          <br />
        </span>
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 220px; height: 60px; --base-color: #2f293a; --highlight-color: #a190fa;"
          >
            ‌
          </span>
          <br />
        </span>
      </div>
      <div
        class="_shopItem_19f3cb"
      >
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 220px; height: 150px; --base-color: #2f293a; --highlight-color: #a190fa;"
          >
            ‌
          </span>
          <br />
        </span>
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 220px; height: 60px; --base-color: #2f293a; --highlight-color: #a190fa;"
          >
            ‌
          </span>
          <br />
        </span>
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 220px; height: 40px; --base-color: #2f293a; --highlight-color: #a190fa;"
          >
            ‌
          </span>
          <br />
        </span>
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 220px; height: 60px; --base-color: #2f293a; --highlight-color: #a190fa;"
          >
            ‌
          </span>
          <br />
        </span>
      </div>
      <div
        class="_shopItem_19f3cb"
      >
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 220px; height: 150px; --base-color: #2f293a; --highlight-color: #a190fa;"
          >
            ‌
          </span>
          <br />
        </span>
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 220px; height: 60px; --base-color: #2f293a; --highlight-color: #a190fa;"
          >
            ‌
          </span>
          <br />
        </span>
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 220px; height: 40px; --base-color: #2f293a; --highlight-color: #a190fa;"
          >
            ‌
          </span>
          <br />
        </span>
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 220px; height: 60px; --base-color: #2f293a; --highlight-color: #a190fa;"
          >
            ‌
          </span>
          <br />
        </span>
      </div>
      <div
        class="_shopItem_19f3cb"
      >
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 220px; height: 150px; --base-color: #2f293a; --highlight-color: #a190fa;"
          >
            ‌
          </span>
          <br />
        </span>
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 220px; height: 60px; --base-color: #2f293a; --highlight-color: #a190fa;"
          >
            ‌
          </span>
          <br />
        </span>
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 220px; height: 40px; --base-color: #2f293a; --highlight-color: #a190fa;"
          >
            ‌
          </span>
          <br />
        </span>
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 220px; height: 60px; --base-color: #2f293a; --highlight-color: #a190fa;"
          >
            ‌
          </span>
          <br />
        </span>
      </div>
      <div
        class="_shopItem_19f3cb"
      >
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 220px; height: 150px; --base-color: #2f293a; --highlight-color: #a190fa;"
          >
            ‌
          </span>
          <br />
        </span>
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 220px; height: 60px; --base-color: #2f293a; --highlight-color: #a190fa;"
          >
            ‌
          </span>
          <br />
        </span>
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 220px; height: 40px; --base-color: #2f293a; --highlight-color: #a190fa;"
          >
            ‌
          </span>
          <br />
        </span>
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 220px; height: 60px; --base-color: #2f293a; --highlight-color: #a190fa;"
          >
            ‌
          </span>
          <br />
        </span>
      </div>
      <div
        class="_shopItem_19f3cb"
      >
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 220px; height: 150px; --base-color: #2f293a; --highlight-color: #a190fa;"
          >
            ‌
          </span>
          <br />
        </span>
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 220px; height: 60px; --base-color: #2f293a; --highlight-color: #a190fa;"
          >
            ‌
          </span>
          <br />
        </span>
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 220px; height: 40px; --base-color: #2f293a; --highlight-color: #a190fa;"
          >
            ‌
          </span>
          <br />
        </span>
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="width: 220px; height: 60px; --base-color: #2f293a; --highlight-color: #a190fa;"
          >
            ‌
          </span>
          <br />
        </span>
      </div>
    </div>
  </div>
</div>
`;
